### ISSUE TITLE
Epetra support through the library.

### ISSUE THREAD

**Issue by bartgol** [gahansen/Albany#222]
This is something that has bugged me for a while now. Albany 'supports' epetra vectors, but does all the work with tpetra vectors under the hood (in particular, an Epetra only build is not supported). This generates a big code duplication, with obvious maintenance problems.

Right now there are people thinking about a linear algebra structures re-design (see issue #198 ). A first look at the work needed for that refactor shows some complications due to Epetra structures, in particular, the fact that the epetra model evaluator is NOT a Thyra model evaluator.

So, let's cut it to the chase: is there a reason to keep the Epetra support at all? I know some things in the optimization framework are only working with Epetra, but @mperego is planning on extending full support to Tpetra. So I guess that probably Epetra should be kept at least until that is done. But what about after that? Are there downstream applications and/or subpackages using explicitly Epetra?

Assuming the answer is 'yes, we need Epetra', then I would like to do some work *before* the Thyra refactor, namely, I would like to push all the E<->T conversion stuff outside of Albany::Application and all the discretization/evaluators structures. Instead, I would try to limit the epetra scope to the ModelEvaluator, which, by the way, I'd like to make inherit from the Thyra adapter to the EpetraExt ME (right now it inherits from the EpetraExt ME). Every call to ModelEvaluator's interfaces would entail some conversion of vectors/matrices. This is not a big deal, since...it is already happening in Albany Application! Perhaps some stuff can be stored in the ME (I'm thinking about storing the maps, so that they do not have to be created every time), but for vectors/matrices, we would do what's done now, except we would do it outside the application. This would free also the discretization from the duplication of code. I also believe the epetra version of the adapt solution manager could then be dropped, but I'm not 100% sure (yet).

So, to recap:

- Do we need Epetra support at all?
- Can we (i.e., any objections?) move Epetra handling out of as many Albany structures as possible?

I think this would make the switch to thyra easier. Although it could be done at the same time as we do that. Not sure which approach would make it easier/faster.

**Comment by ikalash** [comment]
The following physics require Epetra:

- Sophisticated multi-physics QCAD problems (I know we have a P/T to port this fully to Thyra but I do not know when I will have time to do this...); basic QCAD problems should work with Thyra.
- MOR (tagging @plindsa ).

Another thing is ML vs. MueLu preconditioning for FELIX.  I did some testing this year with @tawiesn that revealed that MueLu can be competitive for Antarctica FELIX problem but there seemed to still be some bugs for the finer resolution problems (perhaps you have an update @tawiesn?).  Having the capability to run with ML would be a reason to keep Epetra around.

**Comment by bartgol** [comment]
Thanks @ikalash . Ok, so Epetra should be kept. I guess I have to thing whether cleaning main classes from Epetra *before* switching to Thyra is worth it or whether it should be done as a single step.

**Comment by mperego** [comment]
@bartgol For optimization we currently need the Epetra model evaluator. Anyway, I'm planning to address this first thing in January. It shouldn't take long, hopefully a couple of days.

**Comment by agsalin** [comment]
We discussed this a few weeks ago. I think we should set a date to deprecate Epetra support, and rip it out then. MOR and QCAD and Optimization should be given some time to prepare, either by converting to Tpetra or planning to work off of a branch for a while and take the risk of going stale. 

Maybe April 1, 2018?

**Comment by gahansen** [comment]
We have been having a flurry of discussions on generalizing all (most) Epetra,Tpetra specific instances in the framework to Thyra. Would be an interesting alternative to removing Epetra - and would still support all projects under the Albany "umbrella". Any thoughts or concerns? 

Will continue discussions - specifically with @ikalash @ibaned and @tjfulle

**Comment by bartlettroscoe** [comment]
Thyra does not have interfaces for detailed parallel-aware maps, vector and matrix accessing and loading.  But if you develop your own mini interfaces for exactly the use cases that you need, then this can work.  I think this is basically what they did for Drekar.  You would need to ask Eric C. and Roger P. about that.

**Comment by rppawlo** [comment]
Yes - panzer uses Thyra as an abstraction layer to epetra/tpetra vectors and operators. There are base classes for a linear object factory and a linear object container in panzer that are specialized via inheritance for epetra and tpetra. We have also not been able to pull the trigger on epetra removal due to muelu setup times. @pwxy might have more details/data.

**Comment by ibaned** [comment]
@rppawlo Is there a GitHub issue documenting the MueLu setup time issue? More and more these days I'm hearing about reasons why codes can't switch to MueLu and I'd like to have all of the reasons documented.

**Comment by rppawlo** [comment]
Unfortunately, no. Need to get the latest numbers on Drekar testing from @pwxy. Was there a ticket in nalu for this? Might check with them.

**Comment by bartgol** [comment]
WIP PR #334 is currently addressing the removal of (T/E)petra stuff and its replacement with Thyra objects wherever possible, trying to hide petra-specific stuff into low-level classes (and out of header files), so to keep the high level thyra abstraction intact throughout as much stack calls as possible. If done properly, this would allow to keep the Epetra stack in Albany (provided there are gather/scatter and read/write routines for them). Ideally, this would happen in a single implementation, with a runtime detection of the linear algebra stack type.

I will close this issue for now.
